lib/pure/decimal.nim

```diff
@@ -0,0 +1,1040 @@
+  #     result = left == stripped
+  #   else:
+  #     let stripped = left.strip(-rightExponent)
+  #     result = stripped == right
+  # of dkInfinite:
+  #   result = left == right
+  # of dkNaN:
+  #   result = left == right
+  result = false # TODO
+
+# proc '<'
+
+# proc round  # bankers rounding
+
+# proc roundUp
+
+# proc strip
+
+#
+# assignment / input
+#
+
+proc newDecimal*(numberString: string, places: int = noValue): Decimal =
+  var (decimalKind, negativeFlag, significand, exponent) = parseFromString(numberString)
+  case decimalKind:
+  of dkNaN:
+    result = nan
+  of dkInfinite:
+    if negativeFlag:
+      result = negativeInfinity
+    else:
+      result = infinity
+  of dkValued:
+    result = createDecimal(negativeFlag, significand, exponent)
+    result.places = places
+
+proc `'m`*(numberString: string, places: int = noValue): Decimal =
+  newDecimal(numberString, places)
+
+proc decodeHex*(hex: string): Decimal =
+  if len(hex) != 32:
+    raise newException(ValueError, "The Decimal data type takes 32 hex characters. This string has $1.".format(hex.len))
+  var part = hex[0 .. 7]
+  result.a = fromHex[uint32](part)
+  part = hex[8 .. 15]
+  result.b = fromHex[uint32](part)
+  part = hex[16 .. 23]
+  result.c = fromHex[uint32](part)
+  part = hex[24 .. 31]
+  result.d = fromHex[uint32](part)
+
+#
+# output
+#
+
+proc simpleDigitStr(dList: SignificandArray): string =
+  var firstDigitSeen = false
+  for digit in dList:
+    if digit != 0.byte:
+      firstDigitSeen = true
+    if firstDigitSeen:
+      result &= $(digit.int)
+  if not firstDigitSeen:
+    result = "0"
+
+proc generateNumberString(number: Decimal, exponent: int16, upper: uint32): string =
+  # from the lower three uint32 in decimal and the masked upper uint32 (upper),
+  # derive a sequence of decimal bytes.
+  if number.negative:
+    result = "-"
+  else:
+    result = ""
+
+  let digits = generateDigits(number, upper)
+  let justDigits = simpleDigitStr(digits)
+
+  let scientificExponent = justDigits.len - 1 + exponent
+  if (scientificExponent < -6) or (exponent > 0):
+    # express with scientific notation
+    for index, ch in justDigits:
+      if index == 1:
+        result &= "."
+      result &= ch
+    result &= "E"
+    if scientificExponent >= 0:
+      result &= "+"
+    result &= $scientificExponent
+  elif exponent == 0:
+    # if zero decimal places, then it is a simple integer
+    result &= justDigits
+  else:
+    let significance = -exponent
+    let leadingZeroes = significance - justDigits.len
+    if leadingZeroes >= 0:
+      result &= "0."
+      result &= "0".repeat(leadingZeroes)
+      result &= justDigits
+    else:
+      let depth = -leadingZeroes
+      for index, ch in justDigits:
+        if index == depth:
+          result &= "."
+        result &= ch
+
+proc sci*(number: Decimal): string =
+  ## Express the Decimal value in Scientific Notation
+  var (decKind, exponent, upper) = determineKindExponentAndUpper(number)
```

Comment:
```suggestion
  let (decKind, exponent, upper) = determineKindExponentAndUpper(number)
```
Avoid `var` when you can just use `let` (ditto for other cases).

Reply:
fixed in next commit